pkg/expression/builtin_time.go

```diff
@@ -809,6 +811,166 @@ func (b *builtinLastMonthSig) evalString(ctx EvalContext, row chunk.Row) (string
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETDatetime, types.ETDatetime, types.ETInt)
+	if err != nil {
+		return nil, err
+	}
+
+	bf.setDecimalAndFlenForDatetime(types.MinFsp)
+	sig := &builtinAddMonthSig{bf}
+	return sig, nil
+}
+
+type builtinAddMonthSig struct {
+	baseBuiltinFunc
+}
+
+func (b *builtinAddMonthSig) Clone() builtinFunc {
+	newSig := &builtinAddMonthSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+// evalTime evals a builtinAddMonthSig.
+func (b *builtinAddMonthSig) evalTime(ctx EvalContext, row chunk.Row) (types.Time, bool, error) {
+	date, isNull, err := b.args[0].EvalTime(ctx, row)
+	if isNull || err != nil {
+		return types.ZeroTime, isNull, handleInvalidTimeError(ctx, err)
+	}
+	month, isNull, err := b.args[1].EvalInt(ctx, row)
+	if isNull || err != nil {
+		return types.ZeroTime, isNull, handleInvalidTimeError(ctx, err)
+	}
+
+	orgTime, err := date.GoTime(time.UTC)
+	if err != nil {
+		return types.ZeroTime, true, handleInvalidTimeError(ctx, err)
+	}
+
+	// AddDate() call should have already handle the end of month behavior consistency for us
+	resultTime, err := types.AddDate(0, month, 0, orgTime)
+	if err != nil {
+		return types.ZeroTime, true, handleInvalidTimeError(ctx, types.ErrDatetimeFunctionOverflow.GenWithStackByArgs("datetime"))
+	}
+
+	// in case of datetime overflow
+	if resultTime.Year() == 0 {
+		hour, minute, second := resultTime.Clock()
+		date.SetCoreTime(types.FromDate(0, 0, 0, hour, minute, second, resultTime.Nanosecond()/1000))
+		return date, false, nil
+	}
+
+	var result types.Time
+	result.SetCoreTime(types.FromGoTime(resultTime))
+	tc := typeCtx(ctx)
+	overflow, err := types.DateTimeIsOverflow(tc, date)
+	if err := handleInvalidTimeError(ctx, err); err != nil {
+		return types.ZeroTime, true, err
+	}
+	if overflow {
+		return types.ZeroTime, true, handleInvalidTimeError(ctx, types.ErrDatetimeFunctionOverflow.GenWithStackByArgs("datetime"))
+	}
+
+	return result, false, nil
+}
+
+type nextDayFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *nextDayFunctionClass) getFunction(ctx BuildContext, args []Expression) (builtinFunc, error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETDatetime, types.ETDatetime, types.ETString)
+	if err != nil {
+		return nil, err
+	}
+
+	bf.setDecimalAndFlenForDatetime(types.MinFsp)
+	sig := &builtinNextDaySig{bf}
+	return sig, nil
+}
+
+type builtinNextDaySig struct {
+	baseBuiltinFunc
+}
+
+func (b *builtinNextDaySig) Clone() builtinFunc {
+	newSig := &builtinNextDaySig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+// evalTime evals a builtinNextDaySig.
+func (b *builtinNextDaySig) evalTime(ctx EvalContext, row chunk.Row) (types.Time, bool, error) {
+	// need a map from string to actual weekdays
+	dayMap := map[string]time.Weekday{
```

Comment:
It can be declared as a global variable

Reply:
I think `dayMap` would be better as a local variable here. Currently, I don’t see any other usage for this weekday mapping. If it’s needed elsewhere in the future, we could create a more generalized version and potentially include it in coreTime. For now, `dayMap` also serves as an input check, as the `next_day` function only allows these two types of input, which may not necessarily apply to other functions.